meta/meta.go

```diff
@@ -299,6 +339,33 @@ func (m *Meta) checkTableNotExists(dbKey []byte, tableKey []byte) error {
 	return errors.Trace(err)
 }
 
+func (m *Meta) CreatePolicy(policy *placement.Policy) error {
+	policyKey := m.policyKey(policy.ID)
+
+	if err := m.checkPolicyNotExists(policyKey); err != nil {
+		return errors.Trace(err)
+	}
+	data, err := json.Marshal(policy)
```

Comment:
Just a comment, we could add one byte or uint32 before the json data, as "revision number" or "format version" or "format magic number". It could solve all upgrade/downgrade compatibility problems in the feature.

By checking the head bytes, we could use different parsing methods to unmarshal, or even give different responses. E.g. older TiDB could report error if `GetPolicy` meets a newer version number that only exists on the new TiDB, or do whatever it wants.

Of course the number could be embedded in the json, too. But then you will fail to unmarshal if the persisted data does not match your struct. And you propabably will need to unmarshal it again. In short, you will need to test the version first to give correct reactions.

Reply:
```
func (m *Meta) CheckPolicyHead(value []byte) error {
	var magicVer uint32
	err := json.Unmarshal(value, &magicVer)
	if err != nil {
		return errors.Trace(err)
	}
	if magicVer != placement.CURRENT_POLICY_MAGIC_VERSION {
		return errors.Trace(errors.New("version compatibility error"))
	}
	return nil
}
```

sounds something like the above, once old tidb meets a magic number error, it will errors cause it doesn't know the new structure.

Actually, now by default, we only append something to the structure rather than modifying the old one, so the old tidb can always marshal the value out leaving those appended fields unused.  (it's kind of smooth compatibility). If use magic number above, the error will happen in this case.

Reply:
may be I got your point, the magic version is reserved not for old tidb but the new one.  Future one can change it and check it once there are some new customized handling. Anyway, I added it.

Reply:
>   If use magic number above, the error will happen in this case. It will errors cause it doesn't know the new structure.

Nope, it depends on you. You could say `from version 0x100 - 0x200`, the format will be json, i.e. `0x101-0x199` are minor revisions. Then the old one could "leaving those appended fields unused".

> Future one can change it and check it once there are some new customized handling.

Yes. Not only the old one could choose to reject or partiall handle. The new one could also choose how to handle.

But I actually mean `append(magicBytes, jsonData...)`. A pure binary version number. Then you won't meet such case:

```
var struct MyStruct
if err := json.Unmarshal(value, &struct); err != nil {
}
if struct.magicVer != expectedVersion {
  var struct MyNewStruct
  // OK, we need another older struct definition, some fields are removed in the newer struct definition
  if err := json.Unmarshal(value, &struct); err != nil {
  }
}
```

You will need to unmarshal once to get the version, then can you choose the correct handler. But with a binary number, you could

```
version := getVersion(data[0:xx])
switch version {
case NewVer: // data[xx:] json
case OldVer: // data[xx:] msgpack
case ...: // whatever other formats or meanings
default: // unsupported
}
```

Reply:
Not a necessary request, `magicVersion` field is enough in most cases. You could change it if you like the idea.

Reply:
got it, done